packages/bun-usockets/src/udp.c

```diff
@@ -64,45 +66,67 @@ struct us_udp_packet_buffer_t *us_create_udp_packet_buffer() {
     return (struct us_udp_packet_buffer_t *) bsd_create_udp_packet_buffer();
 }
 
-struct us_internal_udp_t {
-    struct us_internal_callback_t cb;
-    struct us_udp_packet_buffer_t *receive_buf;
-    void (*data_cb)(struct us_udp_socket_t *, struct us_udp_packet_buffer_t *, int);
-    void (*drain_cb)(struct us_udp_socket_t *);
-    void *user;
-    /* An UDP socket can only ever be bound to one single port regardless of how
-     * many interfaces it may listen to. Therefore we cache the port after creation
-     * and use it to build a proper and full sockaddr_in or sockaddr_in6 for every received packet */
-    int port;
-};
+void us_destroy_udp_packet_buffer(struct us_udp_packet_buffer_t *buffer) {
+    bsd_destroy_udp_packet_buffer(buffer);
+}
 
 int us_udp_socket_bound_port(struct us_udp_socket_t *s) {
-    return ((struct us_internal_udp_t *) s)->port;
+    return ((struct us_udp_socket_t *) s)->port;
 }
 
-/* Internal wrapper, move from here */
-void internal_on_udp_read(struct us_udp_socket_t *s) {
-
-    // lookup receive buffer and callback here
-    struct us_internal_udp_t *udp = (struct us_internal_udp_t *) s;
+void us_udp_socket_bound_ip(struct us_udp_socket_t *s, char *buf, int *length) {
+  struct bsd_addr_t addr;
+  if (bsd_local_addr(us_poll_fd((struct us_poll_t *)s), &addr) || *length < bsd_addr_get_ip_length(&addr)) {
+    *length = 0;
+  } else {
+    *length = bsd_addr_get_ip_length(&addr);
+    memcpy(buf, bsd_addr_get_ip(&addr), *length);
+  }
+}
 
-    int packets = us_udp_socket_receive(s, udp->receive_buf);
-    //printf("Packets: %d\n", packets);
+void us_udp_socket_remote_ip(struct us_udp_socket_t *s, char *buf, int *length) {
+  struct bsd_addr_t addr;
+  if (bsd_remote_addr(us_poll_fd((struct us_poll_t *)s), &addr) || *length < bsd_addr_get_ip_length(&addr)) {
+    *length = 0;
+  } else {
+    *length = bsd_addr_get_ip_length(&addr);
+    memcpy(buf, bsd_addr_get_ip(&addr), *length);
+  }
+}
 
-    // we need to get the socket data and lookup its callback here
+void *us_udp_socket_user(struct us_udp_socket_t *s) {
+    struct us_udp_socket_t *udp = (struct us_udp_socket_t *) s;
 
+    return udp->user;
+}
 
-    udp->data_cb(s, udp->receive_buf, packets);
+void us_udp_socket_close(struct us_udp_socket_t *s) {
+  struct us_poll_t *p = (struct us_poll_t *) s;
+//   us_poll_stop(p, s->loop);
+//   bsd_close_socket(us_poll_fd(p));
+    s->closed = 1;
+    // change to writable poll to trigger cleanup on next loop cycle
+    us_poll_change(p, s->loop, LIBUS_SOCKET_WRITABLE);
```

Comment:
what happens if the socket isn't writable until many past ticks? we would be leaving the packet buffer allocated (expensive) and not actually closing the socket, leaking the file descriptor.

IMO, this should:
1) shutdown() the socket so that no more data can be read/written
2) free the packet buffer
3) have a `us_udp_socket* next_to_close` pointer (or similar) and a `us_udp_socket* udp_socket_to_close_head` pointer on the `us_loop_t` struct. then in `us_loop_integrate`, it would check if there are any to close and close them there

Reply:
then, `us_udp_socket_close` could check if it's already closed by seeing if the pointer is not null